### PR TITLE
[LMN-19] Swiftui Carousel

### PR DESCRIPTION
**Pull request by frugoman**
Introducing the Carousel for SwiftUI.

Due to some limitations and hacks required to support infinite scrolling, the approach taken was to move the existing UIKit Carousel component, move it to Backpack-Common, keep the existing Backpack (UIKit) API that is not a facade for the Common component, and in SwiftUI, wrapping that component with a UIViewRepresentable to consume it.

### CODE REVIEW

File: Backpack-Common/Carousel/Classes/BPKInternalCarousel.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+import Foundation
+import UIKit
+
+public protocol BPKInternalCarouselDelegate: AnyObject {
+    func onImageChange(index: Int)
+}
+
+public class BPKInternalCarousel: UIView {
```
Comment (frugoman) [inline Backpack-Common/Carousel/Classes/BPKInternalCarousel.swift:26]: Moving the existing UIKit carousel to Backpack-Common, renaming it to `Internal`.
It still needs to be public, as this is essentially a different module, that needs to be consumed by Backpack and Backpack-SwiftUI.

File: Backpack-SwiftUI/Carousel/Classes/BPKCarousel.swift [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        pageIndicatorView.translatesAutoresizingMaskIntoConstraints = false
+        pageIndicatorView.backgroundColor = .clear
+        let carousel = BPKInternalCarousel(pageIndicator: pageIndicatorView)
+        carousel.delegate = context.coordinator
+        
+        let uiImages = images.map { UIHostingController(rootView: $0).view! }
+        carousel.set(images: uiImages)
+        carousel.setCurrentImage(index: currentIndex)
+        return carousel
+    }
```
Comment (frugoman) [inline Backpack-SwiftUI/Carousel/Classes/BPKCarousel.swift:61]: Wrapping the (now) BPKCommon UIKit Carousel, and injecting it a SwiftUI PageIndicator! Pretty neat!!

File: Backpack-SwiftUI/PageIndicator/Classes/BPKPageIndicator.swift [hunk 3]
```diff
@@ -43,6 +43,7 @@ public struct BPKPageIndicator: UIViewRepresentable {
     
     public func updateUIView(_ uiView: UIPageControl, context: Context) {
         uiView.isUserInteractionEnabled = false
+        uiView.hidesForSinglePage = true
```
Comment (frugoman) [inline Backpack-SwiftUI/PageIndicator/Classes/BPKPageIndicator.swift:46]: The SwiftUI PageIndicator was missing this, where in the UIKit one we do have.
This makes the indicator disappear when there's only on item inside (intended behaviour!)

File: Backpack/Carousel/Classes/BPKCarousel.swift [hunk 4]
```diff
@@ -17,41 +17,40 @@
  */
 
 import Foundation
+import Backpack_Common
 
 public protocol BPKCarouselDelegate: AnyObject {
     func onImageChange(index: Int)
 }
 
 public final class BPKCarousel: UIView {
```
Comment (frugoman) [inline Backpack/Carousel/Classes/BPKCarousel.swift:26]: This is now essentially just calling the internal carousel component and handling the PageIndicator, same as the SwiftUI one!

File: Backpack/Tests/SnapshotTests/BPKCarouselSnapshotTest.swift [hunk 5]
```diff
@@ -27,7 +27,7 @@ class BPKCarouselSnapshotTest: XCTestCase {
     
     override func setUp() {
         super.setUp()
-        isRecording = false
+        isRecording = true
```
Comment (gert-janvercauteren) [inline Backpack/Tests/SnapshotTests/BPKCarouselSnapshotTest.swift:30]: This needs to go back to false

File: Backpack/Tests/SnapshotTests/__Snapshots__/BPKCarouselSnapshotTest/testViewSnapshotWithSecondSelected.dark-mode.png
Comment (gert-janvercauteren) [inline Backpack/Tests/SnapshotTests/__Snapshots__/BPKCarouselSnapshotTest/testViewSnapshotWithSecondSelected.dark-mode.png:1]: The second is not selected on this image

**Review by gert-janvercauteren — COMMENTED** [review]
Left some small comments. 
It seems that index selection no longer works on the component, all images have changed with the first dot active, where the test indicates 2 or 3 to be active. 

There's also a 'recording' = true left in

File: Backpack/Tests/SnapshotTests/BPKCarouselSnapshotTest.swift (hunk 5 above)
Comment (frugoman) [reply Backpack/Tests/SnapshotTests/BPKCarouselSnapshotTest.swift:30]: :O

File: Backpack-SwiftUI/Tests/Carousel/BPKCarouselTests.swift [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import XCTest
+import SwiftUI
+@testable import Backpack_SwiftUI
+
+class BPKCarouselTests: XCTestCase {
+    private func createImage(_ name: String) -> some View {
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/Tests/Carousel/BPKCarouselTests.swift:24]: We probably want to change this func, the param no longer makes sense :D

**Review by gert-janvercauteren — COMMENTED** [review]
Fixed the image snapshot. the issue seems to be with the color profile. 
Can you fix-up the test? Then it's ready to merge :D

**Review by gert-janvercauteren — APPROVED** [review]

**Pull request merged**